src/durable-cache/src/lib.rs

```diff
@@ -293,6 +300,43 @@ impl<C: DurableCacheCodec> DurableCache<C> {
             }
         }
     }
+
+    async fn compare_and_append<I>(
+        &mut self,
+        updates: I,
+        expected_upper: u64,
+        new_upper: u64,
+    ) -> Result<(), UpperMismatch<u64>>
+    where
+        I: IntoIterator<Item = ((C::KeyCodec, C::ValCodec), u64, i64)>,
+    {
+        self.write
+            .compare_and_append(
+                updates,
+                Antichain::from_elem(expected_upper),
+                Antichain::from_elem(new_upper),
+            )
+            .await
+            .expect("usage should be valid")?;
+
+        // Lag the shard's upper by 1 to keep it readable.
+        let downgrade_to = Antichain::from_elem(new_upper.saturating_sub(1));
+
+        // The since handle gives us the ability to fence out other downgraders using an opaque token.
+        // (See the method documentation for details.)
+        // That's not needed here, so we use the since handle's opaque token to avoid any comparison
+        // failures.
+        let opaque = *self.since_handle.opaque();
+        let downgrade = self
+            .since_handle
+            .maybe_compare_and_downgrade_since(&opaque, (&opaque, &downgrade_to))
```

Comment:
How often do we expect new data to be written to the cache? If it's not proportional to time passing (like all the epoch millis collections are), you might want to make this the non-maybe version so we get more prompt compaction.

Reply:
It is not proportional to time passing. There will be one large write on startup and one small write per DDL statement. I've updated to remove `maybe`.